simd-support/simd-r5v.h

```diff
@@ -0,0 +1,473 @@
+	//Vint vnotone = TYPEINT(vmv_v_x)(DS(~1ull,~1), 2*VL);
+	Vint hidx = TYPEINT(vand_vx)(idx, DS(~1ull,~1), 2*VL); // (0, 0, 2, 2, ...)
+	return TYPE(vrgather_vv)(x, hidx, 2*VL);
+}
+static inline V VDUPH(const V x) {
+	Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+	//Vint vone = TYPEINT(vmv_v_x)(1, 2*VL);
+	Vint hidx = TYPEINT(vor_vx)(idx, 1, 2*VL); // (1, 1, 3, 3, ...)
+	return TYPE(vrgather_vv)(x, hidx, 2*VL);
+}
+
+static inline V FLIP_RI(const V x) {
+	Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+	//Vint vone = TYPEINT(vmv_v_x)(1, 2*VL);
+	Vint hidx = TYPEINT(vxor_vx)(idx, 1, 2*VL); // (1, 0, 3, 2, ...)
+	return TYPE(vrgather_vv)(x, hidx, 2*VL);
+}
+
+#define VNEG(a)   TYPE(vfsgnjn_vv)(a,a,2*VL)
+#define VADD(a,b) TYPE(vfadd_vv)(a,b,2*VL)
+#define VSUB(a,b) TYPE(vfsub_vv)(a,b,2*VL)
+#define VMUL(a,b) TYPE(vfmul_vv)(a,b,2*VL)
+#define VFMA(a, b, c)  TYPE(vfmacc_vv)(c,a,b,2*VL) // or vfmadd
+#define VFMS(a, b, c)  TYPE(vfmsac_vv)(c,a,b,2*VL)
+#define VFNMS(a, b, c) TYPE(vfnmsac_vv)(c,a,b,2*VL)  // or vfnmsub
+#define VFMAI(b, c) VADD(c, VBYI(b)) // fixme: improve
+#define VFNMSI(b, c) VSUB(c, VBYI(b)) // fixme: improve
+#define VFMACONJ(b,c)  VADD(VCONJ(b),c) // fixme: improve
+#define VFMSCONJ(b,c)  VFMACONJ(b,VNEG(c)) // fixme: improve
+#define VFNMSCONJ(b,c) VNEG(VFMSCONJ(b,c)) // fixme: improve
+
+#if 1
+static inline V VCONJ(const V x) {
+	Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+	//Vint vone = TYPEINT(vmv_v_x)(1, 2*VL);
+	Vint hidx = TYPEINT(vand_vx)(idx, 1, 2*VL); // (0, 1, 0, 1, 0, 1)
+	//return TYPEMASK(vfsgnjn_vv)(x, x, x, INT2MASK(hidx), 2*VL);
+	return TYPEMERGEDMASK(vfsgnjn_vv)(INT2MASK(hidx), x, x, x, 2*VL);
+}
+#else
+static inline V VCONJ(const V x)
+{
+	Vint partr = VPARTSPLIT; // (all 1, 0, all 1, 0, ...)
+	V xl = TYPEINTERPRETU2F(vreinterpret_v)(TYPEINT(vand_vv)(TYPEINTERPRETF2U(vreinterpret_v)(x), partr, 2*VL)); // set odd elements to 0
+	Vint parti = TYPEINT(vnot_v)(partr, 2*VL); // (0, all 1, 0, all 1, ...)
+	V xh = TYPEINTERPRETU2F(vreinterpret_v)(TYPEINT(vand_vv)(TYPEINTERPRETF2U(vreinterpret_v)(x), parti, 2*VL)); // set even elements to 0
+	return VADD(xl, VNEG(xh));
+}
+#endif
+
+/* can probably be done better */
+static inline V VBYI(V x)
+{
+  x = VCONJ(x);
+  x = FLIP_RI(x);
+  return x;
+}
+
+#if 1
+static inline V VZMUL(V tx, V sr)
+{
+    V tr = VDUPL(tx);
+    V ti = VDUPH(tx);
+    tr = VMUL(sr, tr);
+    sr = VBYI(sr);
+    return VFMA(ti, sr, tr);
+}
+
+static inline V VZMULJ(V tx, V sr)
+{
+    V tr = VDUPL(tx);
+    V ti = VDUPH(tx);
+    tr = VMUL(sr, tr);
+    sr = VBYI(sr);
+    return VFNMS(ti, sr, tr);
+}
+
+static inline V VZMULI(V tx, V sr)
+{
+    V tr = VDUPL(tx);
+    V ti = VDUPH(tx);
+    ti = VMUL(ti, sr);
+    sr = VBYI(sr);
+    return VFMS(tr, sr, ti);
+}
+
+static inline V VZMULIJ(V tx, V sr)
+{
+    V tr = VDUPL(tx);
+    V ti = VDUPH(tx);
+    ti = VMUL(ti, sr);
+    sr = VBYI(sr);
+    return VFMA(tr, sr, ti);
+}
+#else
```

Comment:
VZMULJ, VZMULI, and VZMULIJ in else macro are the same as them in if macro, so they may be coalesced